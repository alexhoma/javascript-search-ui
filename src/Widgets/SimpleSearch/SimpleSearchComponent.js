/**
 * @jsx h
 */

import { h, Component } from 'preact';
import {simpleSearchAction} from "./simpleSearchActions";

/**
 * SimpleSearch Component
 */
class SimpleSearchComponent extends Component {
    shouldComponentUpdate() {
        return false;
    }

    handleSearch = (e) => {
        const {
            startSearchOn,
            environmentId,
            currentQuery,
            client
        } = this.props;

        /**
         * Search when string is bigger than {startSearchOn}
         */
        if (e.target.value.length < startSearchOn) return;

        /**
         * Dispatch input search action
         */
        simpleSearchAction(
            {   // queryOptions
                queryText: e.target.value
            },
            {   // appOptions
                environmentId,
                currentQuery,
                client
            }
        )
    };

    render() {
        const {
            placeholder,
            autofocus,
            classNames: {
                container: containerClassName,
                input: inputClassName
            }
        } = this.props;

        return (
            <div className={`asui-simpleSearch ${containerClassName}`}>
                <input
                    type='text'
                    className={`asui-simpleSearch--input ${inputClassName}`}
                    autofocus={autofocus}
                    placeholder={placeholder}
                    onInput={this.handleSearch}
                />
            </div>
        );
    }
}

SimpleSearchComponent.defaultProps = {
    placeholder: '',
    autofocus: false,
    startSearchOn: 0,
    classNames: {
        container: '',
        input: ''
    }
};

export default SimpleSearchComponent;