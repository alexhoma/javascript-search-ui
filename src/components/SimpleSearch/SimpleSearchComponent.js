/**
 * @jsx h
 */

import { h, Component } from 'preact';
import {simpleSearchAction} from "./simpleSearchActions";
import Template from "../Template";

/**
 * SimpleSearch Component
 */
class SimpleSearchComponent extends Component {

    handleSearch = (e) => {
        const {
            startSearchOn,
            environmentId,
            currentQuery,
            client
        } = this.props;

        /**
         * Search when string is bigger than {startSearchOn}
         */
        if (e.target.value.length < startSearchOn) return;

        /**
         * Dispatch input search action
         */
        simpleSearchAction(
            {
                queryText: e.target.value
            },
            {
                environmentId,
                currentQuery,
                client
            }
        )
    };

    clearSearch = () => {
        const {
            environmentId,
            currentQuery,
            client
        } = this.props;

        simpleSearchAction(
            {
                queryText: ''
            },
            {
                environmentId,
                currentQuery,
                client
            }
        )
    };

    render() {
        const {
            placeholder,
            autofocus,
            clearSearch,
            withContainer,
            classNames: {
                container: containerClassName,
                input: inputClassName,
                clearSearch: clearSearchClassName
            },
            template: {
                clearSearch: clearSearchTemplate
            },
            currentQuery: {
                q: currentQueryText
            },
            htmlNodeInheritProps
        } = this.props;

        let searchInput = (<input
            type='text'
            className={`as-simpleSearch__input ${inputClassName}`}
            {...htmlNodeInheritProps}
            autofocus={autofocus}
            placeholder={placeholder}
            onInput={this.handleSearch}
            value={currentQueryText}
            autocomplete="off"
        />);

        if (withContainer) {
            return (
                <div className={`as-simpleSearch ${containerClassName}`}>
                    {searchInput}

                    {(clearSearch && currentQueryText.length !== 0)
                        ? (
                            <div
                                className={`as-simpleSearch__clearSearch ${clearSearchClassName}`}
                                onClick={this.clearSearch}
                            >
                                <Template template={clearSearchTemplate}/>
                            </div>
                        ) : null
                    }
                </div>
            );
        }

        return searchInput;
    }
}

SimpleSearchComponent.defaultProps = {
    placeholder: '',
    autofocus: false,
    startSearchOn: 0,
    clearSearch: true,
    withContainer: true,
    classNames: {
        container: '',
        input: '',
        clearSearch: ''
    },
    template: {
        clearSearch: 'x'
    }
};

export default SimpleSearchComponent;