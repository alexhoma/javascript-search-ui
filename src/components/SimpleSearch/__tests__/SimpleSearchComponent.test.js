/** @jsx h */
import {h} from 'preact';
import {shallow, deep} from 'preact-render-spy';
import SimpleSearchComponent from './../SimpleSearchComponent';

describe('<SimpleSearchComponent />', () => {
    const defaultProps = {
        target: '#some-tagret',
        currentQuery: {
            q: ''
        }
    };

    it('should render with default props', () => {
        const tree = shallow(
            <SimpleSearchComponent {...defaultProps} />
        );

        expect(tree).toMatchSnapshot();
    });
    it('should render with custom props', () => {
        /**
         * Full custom props available object
         */
        const customProps = {
            placeholder: 'Type something...',
            classNames: {
                container: 'custom-container-class'
            }
        };

        const tree = shallow(
            <SimpleSearchComponent
                {...defaultProps}
                {...customProps}
            />
        );

        expect(tree).toMatchSnapshot();
    });
    it('should render only the input when "withContainer" property is set to false', () => {
        const tree = shallow(
            <SimpleSearchComponent
                {...defaultProps}
                withContainer={false}
                clearSearch={false}
            />
        );

        expect(tree).toMatchSnapshot();
    });
    it('should render new html properties when "htmlNodeInheritProps" are passed', () => {
        /**
         * This is due when the targeted node is an input
         * instead of a container. When the system takes the selected "input"
         * gets its current properties and replaces this input with the
         * SimpleSearchComponent input.
         */
        const tree = shallow(
            <SimpleSearchComponent
                {...defaultProps}
                htmlNodeInheritProps={{
                    'className': 'some-class-name',
                    'data-apisearch': 'apisearch-search-bar',
                    'id': 'some-id',
                    'autocomplete': 'off',
                    'spellcheck': 0
                }}
                withContainer={false}
                clearSearch={false}
            />
        );

        expect(tree).toMatchSnapshot();
    });
    it('should handleSearch() on input event', () => {
        const tree = shallow(
            <SimpleSearchComponent {...defaultProps} />
        );

        tree.find('input')[0].attributes.onInput = jest.fn();
        tree.find('input').simulate('input', 'this is a search query text');

        expect(
            tree.find('input').attr('onInput')
        ).toBeCalledWith('this is a search query text')
    });
    it('should appear a clearSearch button when the input value is not empty, and should handle clearSearch() on click', () => {
        /**
         * A text is typed on the input
         */
        const tree = deep(
            <SimpleSearchComponent
                {...defaultProps}
                currentQuery={{q: 'Some query text...'}}
            />
        );
        expect(tree).toMatchSnapshot();

        /**
         * The clearSearch button is clicked
         */
        const trigger = jest.fn();
        tree.find('.as-simpleSearch__clearSearch')[0].attributes.onClick = trigger;
        tree.find('.as-simpleSearch__clearSearch').simulate('click');
        expect(trigger).toBeCalled();
    });
});