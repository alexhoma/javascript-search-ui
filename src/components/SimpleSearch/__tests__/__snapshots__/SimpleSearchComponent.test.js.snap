// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SimpleSearchComponent /> should appear a clearSearch button when the input value is not empty, and should handle clearSearch() on click 1`] = `
preact-render-spy (1 nodes)
-------
<div class="as-simpleSearch ">
  <input
    type="text"
    class="as-simpleSearch__input "
    autofocus={false}
    placeholder=""
    onInput={[Function anonymous]}
    value="Some query text..."
    autocomplete="off"
   />
  <div
    class="as-simpleSearch__clearSearch "
    onClick={[Function anonymous]}
  >
    <div>x</div>
  </div>
</div>

`;

exports[`<SimpleSearchComponent /> should render new html properties when "htmlNodeInheritProps" are passed 1`] = `
preact-render-spy (1 nodes)
-------
<input
  type="text"
  class="some-class-name"
  data-apisearch="apisearch-search-bar"
  id="some-id"
  autofocus={false}
  placeholder=""
  onInput={[Function anonymous]}
  value=""
  autocomplete="off"
 />

`;

exports[`<SimpleSearchComponent /> should render only the input when "withContainer" property is set to false 1`] = `
preact-render-spy (1 nodes)
-------
<input
  type="text"
  class="as-simpleSearch__input "
  autofocus={false}
  placeholder=""
  onInput={[Function anonymous]}
  value=""
  autocomplete="off"
 />

`;

exports[`<SimpleSearchComponent /> should render with custom props 1`] = `
preact-render-spy (1 nodes)
-------
<div class="as-simpleSearch custom-container-class">
  <input
    type="text"
    class="as-simpleSearch__input undefined"
    autofocus={false}
    placeholder="Type something..."
    onInput={[Function anonymous]}
    value=""
    autocomplete="off"
   />
</div>

`;

exports[`<SimpleSearchComponent /> should render with default props 1`] = `
preact-render-spy (1 nodes)
-------
<div class="as-simpleSearch ">
  <input
    type="text"
    class="as-simpleSearch__input "
    autofocus={false}
    placeholder=""
    onInput={[Function anonymous]}
    value=""
    autocomplete="off"
   />
</div>

`;
